File: 002-AddTwoNumbers/C++/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int added = 0, num = 0, carry = 0;
        
        ListNode *mylist = new ListNode(0);
        ListNode *dummy = mylist;
        
        // loop over the two list
        for (; l1 != NULL && l2 != NULL; l1 = l1->next, l2 = l2->next) {
            added = l1->val + l2->val + carry;
            num = added % 10; carry = added / 10;
            
            mylist->next = new ListNode(num);
            mylist = mylist->next;
        }
        
        // loop over one list
        for (; l1 != NULL; l1 = l1->next) {
            added = l1->val + carry;
            num = added % 10; carry = added / 10;
            
            mylist->next = new ListNode(num);
            mylist = mylist->next;
        }
        
        for (; l2 != NULL; l2 = l2->next) {
            added = l2->val + carry;
            num = added % 10; carry = added / 10;
            
            mylist->next = new ListNode(num);
            mylist = mylist->next;
        }
        
        // don't forget the carry
        if (carry != 0) {
            mylist->next = new ListNode(carry);
            mylist = mylist->next;
        }
        
        return dummy->next;
        
    }
};

File: 002-AddTwoNumbers/Go/main.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var added, num, carry int
    
    mylist := &ListNode{0, nil}
    dummy := mylist
    
    // loop over the two list
    for ;l1 != nil && l2 != nil; l1, l2 = l1.Next, l2.Next {
        added = l1.Val + l2.Val + carry
        num, carry = added % 10, added / 10
        
        mylist.Next = &ListNode{num, nil}
        mylist = mylist.Next
    }
    
    // loop over one list
    for ;l1 != nil; l1 = l1.Next {
        added = l1.Val + carry
        num, carry = added % 10, added / 10
        
        mylist.Next = &ListNode{num, nil}
        mylist = mylist.Next
    }
    
    for ;l2 != nil; l2 = l2.Next {
        added = l2.Val + carry
        num, carry = added % 10, added / 10
        
        mylist.Next = &ListNode{num, nil}
        mylist = mylist.Next
    }
    
    // don't forget the carry baby!
    if carry != 0 {
        mylist.Next = &ListNode{carry, nil}
        mylist = mylist.Next
    }
    
    return dummy.Next
    
}